components/brave_today/rust/lib.rs

```diff
@@ -0,0 +1,121 @@
+extern crate feed_rs;
+use feed_rs::parser;
+use lazy_static::lazy_static;
+use regex::Regex;
+
+#[cxx::bridge(namespace = brave_news)]
+mod ffi {
+    pub struct FeedItem {
+        id: String,
+        title: String,
+        description: String,
+        image_url: String,
+        destination_url: String,
+        published_timestamp: i64,
+    }
+
+    pub struct FeedData {
+        id: String,
+        title: String,
+        items: Vec<FeedItem>,
+    }
+
+    extern "Rust" {
+        fn parse_feed_string(source: String, output: &mut FeedData) -> bool;
+    }
+}
+
+fn parse_feed_string(source: String, output: &mut ffi::FeedData) -> bool {
+    lazy_static! {
+        static ref IMAGE_REGEX: Regex = Regex::new("<img[^>]+src=\"([^\">]+)").unwrap();
+    }
+    // TODO(petemill): To see these errors, put `env_logger::init()` in an
+    // 'init' function which only gets called once.
+    let feed_result = parser::parse(source.as_bytes());
+    if let Err(error) = feed_result {
+        log::debug!("Could not parse feed: {}", error);
+        return false;
+    }
+    // Parsing was successful, convert to FeedData
+    let feed = feed_result.unwrap();
+    output.title = if let Some(title) = feed.title {
+        voca_rs::strip::strip_tags(&title.content)
+    } else {
+        String::new()
+    };
+    for feed_item_data in feed.entries {
+        // Check we can make a valid entry
+        if feed_item_data.links.is_empty()
+            || feed_item_data.published.is_none()
+            || (feed_item_data.title.is_none() && feed_item_data.summary.is_none())
+            || feed_item_data.published.is_none()
+        {
+            // Ignore if we can't use this entry
+            continue;
+        }
+        let mut image_url = String::new();
+        // Attempt to get an image by looking at a couple possible fields:
+        //  - media[].content[].url
+        //  - media[].thumbnails[].image.uri
+        for media_object in feed_item_data.media {
+            let mut largest_width = 0;
+            for content_item in media_object.content {
+                if let Some(content_item_url) = content_item.url {
+                    let this_width = content_item.width.unwrap_or(0);
+                    if this_width >= largest_width {
+                        image_url = content_item_url.to_string();
+                        largest_width = this_width;
+                    }
+                }
+            }
+            if !image_url.is_empty() {
+                continue;
```

Comment:
I found the logic a hard to follow here. maybe comment that you're using the longest url from media[].content[], so it's clear why your skipping over the subsequent thumbnail search?